docs/specs/authentication-flow-prioritized-identity.md

```diff
@@ -0,0 +1,268 @@
+# Authentication Flow - Prioritized Identity
+
+- [Introduction](#introduction)
+- [Usecases](#usecases)
+  - [Force user to use a preferred authentication method if available](#force-user-to-use-a-preferred-authentication-method-if-available)
+- [Behaviors](#behaviors)
+  - [When there are multiple levels of priority](#when-there-are-multiple-levels-of-priority)
+  - [When the user has only some of the prioritized identities](#when-the-user-has-only-some-of-the-prioritized-identities)
+## Introduction
+
+It is common that some identities are considered more secure, or more preferable comparing with other identities of the same user. Apps might want to have a way to specify the priority of different types of identities in a login flow, so when a user has another identity which is more preferable, it will be used for identification inside a login flow.
+
+## Usecases
+
+### Force user to use a preferred authentication method if available
+
+Assume you have a portal app which supports two signup methods: Email with Password, and Google oauth login.
+
+User of the portal can signup to the portal by one of the two methods, and login with that method afterwards.
+
+If the user has signed up with email with password, the user can connect to a Google account at any time after he logged in to the portal (For example, through [Account Linking](./account-linking.md)). In your persepctive, Google oauth login is the preferred login method because it is more secure. Therefore once user connected their google account, the portal should only accept google oauth login for the same account, and do not accept email with password logins.
+
+Use the following config to acheive the goal:
+
+```yaml
+authentication_flow:
+  login_flows:
+    - name: default
+      steps:
+        - name: identify
+          type: identify
+          one_of:
+            - identification: oauth
+              priority: 1 # <- Configure priority here
```

Comment:
Adding priority here does not allow me to specify "Login with Google" is preferred over "Login with X". I think we need to allow specifing priority for each alias. A few problems then arise.

1. The developer wants to specify priority for a alias. So she must write out the alias. For aliases that were not written out, what is the meaning?
  - Aliases that are not written out are omitted from this identification. This enables another feature that is not implemented yet.
  - Aliases that are not written out are still included in this identification. Their priority is 0.

Reply:
It seems the problem here is we didn't consider different oauth method as different identification options, but they should be in fact separated options. My suggestion is:

```yaml
steps:
  - type: identify
    one_of:
        - identification: oauth
          priority: 2
          alias: adfs
        - identification: oauth
          priority: 1
          alias: google
        - identification: oauth
          priority: 0
```

1. Add `alias` to `identification: oauth`. If `alias` is set, only provider of that alias will be included as option.
2. If `alias` is not set, it means "all" provider.
3. `priority` applies to the specified oauth provider if `alias` is set, or apply to all providers if `alias` is not set.

Reply:
From your example config, due to the presence of the 3rd item, all OAuth providers are shown as options. The actual number of options could be more than 3. It is not intuitive.

What about

```yaml
steps:
- type: identify
  one_of:
  - identification: oauth
    aliases: # For exposing only a subset of configured OAuth providers.
    - alias: google
    - alias: adfs
    priorities: # Specify the priority
    - alias: google
      priority: 1
  - identification: email
  - identification: phone

```

Reply:
Some questions in you example:
1. How to specify priority for email and phone? I guess it will be something like this:

```yaml
steps:
- type: identify
  one_of:
  - identification: oauth
    aliases: # For exposing only a subset of configured OAuth providers.
    - alias: google
    - alias: adfs
    priorities: # Specify the priority
    - alias: google
      priority: 1
  - identification: email
     priority: 1 # <--
  - identification: phone
```

but this is a bit confusing. The identification object now has two fields, `priorities` which is for oauth, and `priority` for other identifications.

2. How to specify priority for "any provider"?

> From your example config, due to the presence of the 3rd item, all OAuth providers are shown as options. The actual number of options could be more than 3. It is not intuitive.

This is actually a problem exist in the current config. Anyway, for backward compatibility, we should support `identification: oauth` without a specified `aliases`, so this problem is actually unavoidable. In your example, this is also possible:

```yaml
steps:
- type: identify
  one_of:
  - identification: oauth
     priorities: # Specify the priority
    - # Not sure how to write
      priority: 1
  - identification: email
  - identification: phone
```

And the resulting number of options is still not same as the number of rows of `identification`.

Reply:
p.s. right now portal have no way to config these authflow, so if something seriously wrong in our existing design, we can actually change it and no need to consider backward compatibilities (we can always modify all existing authflow config after we change it) -- this is intended. @louischan-oursky @tung2744

Reply:
Then I think if we really think `number of options != number of identification` is a problem, I suggest we simply deprecate `identification: oauth` without `alias`. i.e., we always ask developer to provide a `alias` if `identification: oauth` is used.

And for migration of old data, for simplicity, we can just leave the current behavior of `identification: oauth` without `alias`, but we will stop generating config in this way, and also remove it from documentation.

Reply:
> And for migration of old data, for simplicity, we can just leave the current behavior of identification: oauth without alias, but we will stop generating config in this way, and also remove it from documentation.

Agree. We should keep supporting it.

On the other hand, this deprecated config must not co-exist with the new one. It is a config error if such case happens.